File: briar-headless/src/main/java/org/briarproject/briar/headless/messaging/OutputPrivateMessage.java
package org.briarproject.briar.headless.messaging;

import org.briarproject.bramble.api.nullsafety.NotNullByDefault;
import org.briarproject.briar.api.messaging.PrivateMessage;
import org.briarproject.briar.api.messaging.PrivateMessageHeader;

import javax.annotation.concurrent.Immutable;

@Immutable
@NotNullByDefault
@SuppressWarnings("WeakerAccess")
public class OutputPrivateMessage {

	public final String body;
	public final long timestamp;
	public final boolean read, seen, sent, local;
	public final byte[] id, groupId;

	OutputPrivateMessage(PrivateMessageHeader header, String body) {
		this.body = body;
		this.timestamp = header.getTimestamp();
		this.read = header.isRead();
		this.seen = header.isSeen();
		this.sent = header.isSent();
		this.local = header.isLocal();
		this.id = header.getId().getBytes();
		this.groupId = header.getGroupId().getBytes();
	}

	/**
	 * Only meant for own {@link PrivateMessage}s directly after creation.
	 */
	OutputPrivateMessage(PrivateMessage m, String body) {
		this.body = body;
		this.timestamp = m.getMessage().getTimestamp();
		this.read = true;
		this.seen = true;
		this.sent = false;
		this.local = true;
		this.id = m.getMessage().getId().getBytes();
		this.groupId = m.getMessage().getGroupId().getBytes();
	}
}

File: briar-headless/src/main/java/org/briarproject/briar/headless/messaging/OutputPrivateMessage.kt
package org.briarproject.briar.headless.messaging

import org.briarproject.bramble.api.contact.ContactId
import org.briarproject.briar.api.messaging.PrivateMessage
import org.briarproject.briar.api.messaging.PrivateMessageHeader
import org.briarproject.briar.headless.json.JsonDict

internal fun PrivateMessageHeader.output(contactId: ContactId) = JsonDict(
    "type" to "PrivateMessage",
    "contactId" to contactId.int,
    "timestamp" to timestamp,
    "read" to isRead,
    "seen" to isSeen,
    "sent" to isSent,
    "local" to isLocal,
    "id" to id.bytes,
    "groupId" to groupId.bytes
)

internal fun PrivateMessageHeader.output(contactId: ContactId, text: String?): JsonDict {
    val dict = output(contactId)
    dict["text"] = text
    return dict
}

/**
 * Use only for outgoing messages that were just sent
 */
internal fun PrivateMessage.output(contactId: ContactId, text: String) = JsonDict(
    "type" to "PrivateMessage",
    "contactId" to contactId.int,
    "timestamp" to message.timestamp,
    "read" to true,
    "seen" to false,
    "sent" to false,
    "local" to true,
    "id" to message.id.bytes,
    "groupId" to message.groupId.bytes,
    "text" to text
)
